### DISCUSSION TITLE
Extend vercel.json to allow settings definition in code

### DISCUSSION THREAD

**Discussion by luke-j** [vercel/vercel#6549] [Ideas]
I would like to define the settings for my projects in code. Ideally, I would like to be able to do something like:

```json
{
  "headers": [ … ],
  "rewrites": [ … ],
  "settings": {
    "domains": ["example.com"],
    "git": {
      "productionBranch": "main"
    },
    "security": {
      "forkProtection": true
    }
  }
}
```

The occasional open-source tool pops up to deal with this issue, for example https://github.com/chronark/terraform-provider-vercel.

Are there any plans to extend `vercel.json` to allow this functionality?

**Comment by styfle** [comment]
It wouldn't be safe to modify project settings per deployment.

Imagine this scenario:

1. Set Production branch to `main` in project settings via dashboard (like it works today)
2. Push to `main` branch without vercel.json - deploy to production
3. Push to `feature` branch with `{  "git": { "productionBranch": "feature"  } }` - deploy to production and modify project settings
4. Push to `another` branch without vercel.json - deploy to preview 
5. Merge `another` to `main`, still no vercel.json - deploy to preview (**unexpected behavior**)

**Comment by Ethan-Arrowood** [comment]
Hi @luke-j !

Good news, Vercel is going to be shipping a feature similar to what you have requested here. While not all settings will be supported, we are adding the ability to override certain project settings so you can have per-deployment configuration. We are looking for some customers to try it out before we release it to everyone; if this is of interest to you, please reach out to me (ethan.arrowood@vercel.com) and lets find some time to chat and try it out together! Thanks!
